datafusion/functions-array/src/kernels.rs

```diff
@@ -252,3 +253,67 @@ pub(super) fn array_to_string(args: &[ArrayRef]) -> datafusion_common::Result<Ar
 
     Ok(Arc::new(string_arr))
 }
+
+use arrow::array::ListArray;
+use arrow::buffer::OffsetBuffer;
+use arrow::datatypes::Field;
+/// Generates an array of integers from start to stop with a given step.
+///
+/// This function takes 1 to 3 ArrayRefs as arguments, representing start, stop, and step values.
+/// It returns a `Result<ArrayRef>` representing the resulting ListArray after the operation.
+///
+/// # Arguments
+///
+/// * `args` - An array of 1 to 3 ArrayRefs representing start, stop, and step(step value can not be zero.) values.
+///
+/// # Examples
+///
+/// gen_range(3) => [0, 1, 2]
+/// gen_range(1, 4) => [1, 2, 3]
+/// gen_range(1, 7, 2) => [1, 3, 5]
+pub fn gen_range(
+    args: &[ArrayRef],
+    include_upper: i64,
+) -> datafusion_common::Result<ArrayRef> {
+    let (start_array, stop_array, step_array) = match args.len() {
+        1 => (None, as_int64_array(&args[0])?, None),
+        2 => (
+            Some(as_int64_array(&args[0])?),
+            as_int64_array(&args[1])?,
+            None,
+        ),
+        3 => (
+            Some(as_int64_array(&args[0])?),
+            as_int64_array(&args[1])?,
+            Some(as_int64_array(&args[2])?),
+        ),
+        _ => return exec_err!("gen_range expects 1 to 3 arguments"),
+    };
+
+    let mut values = vec![];
+    let mut offsets = vec![0];
+    for (idx, stop) in stop_array.iter().enumerate() {
+        let stop = stop.unwrap_or(0) + include_upper;
```

Comment:
`generate_series(i64::MAX, i64::MAX)` will panic.

```sh
DataFusion CLI v36.0.0
❯ select generate_series(9223372036854775807, 9223372036854775807);
thread 'main' panicked at datafusion/functions-array/src/kernels.rs:296:20:
attempt to add with overflow
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```
It can succeed in PostgreSQL and DuckDB.

Reply:
I think we don't have i128Array yet, so probably this panic is unavoidable until we support it.

Reply:
And the result is incorrect when the step is a negative number.
```sh
DataFusion CLI v36.0.0
❯ select generate_series(5,1,-1);
+----------------------------------------------+
| generate_series(Int64(5),Int64(1),Int64(-1)) |
+----------------------------------------------+
| [5, 4, 3]                                    |
+----------------------------------------------+
1 row in set. Query took 0.005 seconds.
```

In DuckDB:
```
D select generate_series(5,1,-1);
┌───────────────────────────┐
│ generate_series(5, 1, -1) │
│          int64[]          │
├───────────────────────────┤
│ [5, 4, 3, 2, 1]           │
└───────────────────────────┘
```

Reply:
> I think we don't have i128Array yet, so probably this panic is unavoidable until we support it.

I tried it in the following way, and then it worked, but I haven't checked it carefully yet.
```rust
    for (idx, stop) in stop_array.iter().enumerate() {
        let stop = stop.unwrap_or(0);
        let start = start_array.as_ref().map(|arr| arr.value(idx)).unwrap_or(0);
        let step = step_array.as_ref().map(|arr| arr.value(idx)).unwrap_or(1);
        if step == 0 {
            return exec_err!("step can't be 0 for function range(start [, stop, step]");
        }
        if step < 0 {
            // Decreasing range
            values.extend((stop + 1..start + 1).rev().step_by((-step) as usize));
        } else {
            // Increasing range
            values.extend((start..stop).step_by(step as usize));
        }
        // TODO: include_upper should be a boolean flag
        if include_upper > 0 {
            match values.last() {
                Some(&last) if last + step == stop => {
                    values.push(stop);
                }
                None => {
                    values.push(stop);
                }
                _ => {}
            }
        }
        offsets.push(values.len() as i32);
    }
```

Reply:
Leaving the edge cases for future handling is okay with me. However, the behavior for negative step is incorrect and needs to be fixed.

Reply:
Got it